File: typescript/src/concurrent10.ts
import Computer from 'tzafon';
import path from 'node:path';
import { ensureDir, timestamp, downloadToFile } from './utils.js';
import readline from 'node:readline/promises';
import { stdin as input, stdout as output } from 'node:process';

const URLS: [string, string][] = [
  ['wikipedia', 'https://www.wikipedia.org/'],
  ['nytimes', 'https://www.nytimes.com/'],
  ['airbnb', 'https://www.airbnb.com/'],
  ['github', 'https://github.com/'],
  ['reddit', 'https://www.reddit.com/']
];

async function worker(i: number, label: string, url: string, total?: number) {
  const outDir = path.join(path.dirname(new URL(import.meta.url).pathname), `../results/concurrent_10_${label}`);
  ensureDir(outDir);
  const client = new (Computer as any)();
  const computer = await client.create({ kind: 'browser' });
  await computer.navigate(url);
  try { await computer.wait(2); } catch {}
  const result = await computer.screenshot();
  const shotUrl = (result as any)?.result?.screenshot_url as string | undefined;
  if (shotUrl) {
    const img = path.join(outDir, `${timestamp(`${label}_${i}_`)}.png`);
    await downloadToFile(shotUrl, img);
    const prefix = total ? `[${i + 1}/${total}]` : `[${i + 1}]`;
    console.log(`${prefix} Saved: ${img}`);
    return img;
  }
  return '';
}

function parseArgsSite(): string | undefined {
  const idx = process.argv.indexOf('--site');
  if (idx >= 0 && process.argv[idx + 1]) return process.argv[idx + 1];
  return undefined;
}

function pickLabelUrl(site?: string): [string, string] {
  const labels = new Map(URLS);
  if (site) {
    if (labels.has(site)) return [site, labels.get(site)!];
    if (site.startsWith('http://') || site.startsWith('https://')) {
      try {
        const u = new URL(site);
        const host = u.hostname;
        const label = host.includes('.') ? host.split('.').slice(-2, -1)[0] : host;
        return [label || 'site', site];
      } catch { /* fallthrough */ }
      return ['site', site];
    }
  }
  return ['',''];
}

async function chooseSiteInteractive(): Promise<[string, string]> {
  const rl = readline.createInterface({ input, output });
  try {
    console.log('Choose a site to screenshot concurrently (10 shots):');
    URLS.forEach(([lbl, url], i) => console.log(`  ${i + 1}. ${lbl} -> ${url}`));
    const ans = await rl.question(`Enter number (1-${URLS.length}): `);
    const idx = Math.min(Math.max(parseInt(ans || '1', 10) || 1, 1), URLS.length) - 1;
    return URLS[idx];
  } finally {
    rl.close();
  }
}

async function run(n = 10) {
  let [label, url] = pickLabelUrl(parseArgsSite());
  if (!label || !url) {
    [label, url] = await chooseSiteInteractive();
  }
  const tasks = Array.from({ length: n }, (_, i) => worker(i, label, url, n));
  const results = await Promise.allSettled(tasks);
  for (const r of results) {
    if (r.status === 'fulfilled') console.log(`Saved: ${r.value}`);
    else console.warn('Worker failed:', r.reason);
  }
}

run(10);

File: python/concurrent_10.py
import os
import random
import argparse
import time
import asyncio
import traceback
from typing import List, Tuple, Dict

from tzafon import Computer
from utils import ensure_dir, timestamp, download_to_file


URLS = [
    ("wikipedia", "https://www.wikipedia.org/"),
    ("nytimes", "https://www.nytimes.com/"),
    ("airbnb", "https://www.airbnb.com/"),
    ("github", "https://github.com/"),
    ("reddit", "https://www.reddit.com/"),
]


def _select_site_arg(site: str | None) -> Tuple[str, str]:
    label_to_url: Dict[str, str] = {k: v for k, v in URLS}
    if site:
        if site in label_to_url:
            return site, label_to_url[site]
        if site.startswith("http://") or site.startswith("https://"):
            from urllib.parse import urlparse
            try:
                host = urlparse(site).hostname or "site"
                label = host.split(".")[-2] if "." in host else host
            except Exception:
                label = "site"
            return label, site

    print("Choose a site to screenshot concurrently (10 shots):")
    for idx, (lbl, url) in enumerate(URLS, start=1):
        print(f"  {idx}. {lbl} -> {url}")
    choice = input("Enter number (1-{}): ".format(len(URLS))).strip()
    try:
        i = int(choice)
        assert 1 <= i <= len(URLS)
    except Exception:
        i = 1
    return URLS[i - 1]


def _cleanup(client: object, c: object) -> None:
    for obj in (c, client):
        for name in (
            "close", "quit", "shutdown", "destroy", "stop", "end", "exit", "delete", "dispose"
        ):
            try:
                fn = getattr(obj, name, None)
                if callable(fn):
                    fn()
            except Exception:
                pass

    try:
        delete = getattr(client, "delete", None)
        if callable(delete):
            try:
                delete(c)
            except Exception:
                comp_id = getattr(c, "id", None)
                if comp_id is not None:
                    try:
                        delete(comp_id)
                    except Exception:
                        pass
    except Exception:
        pass


def _create_browser_with_retry(client: object, retries: int = 6) -> object:
    delay = 2.0
    for attempt in range(retries):
        try:
            return client.create(kind="browser")
        except Exception as e:
            msg = str(e).lower()
            if "429" in msg or "concurrent" in msg or "limit" in msg:
                time.sleep(delay)
                delay = min(delay * 2, 30.0)
                continue
            raise
    return client.create(kind="browser")


def _is_capacity_error(e: Exception) -> bool:
    msg = str(e).lower()
    return ("429" in msg) or ("concurrent" in msg and "computer" in msg) or ("limit" in msg)


def _take_one(i: int, label: str, url: str, client: object | None = None) -> str:
    base = os.path.join(os.path.dirname(__file__), "results", f"concurrent_10_{label}")
    ensure_dir(base)
    client = client or Computer()
    c = _create_browser_with_retry(client)
    c.navigate(url)
    try:
        c.wait(2)
    except Exception:
        pass
    try:
        result = c.screenshot()
        try:
            shot_url = result.result["screenshot_url"]
        except Exception:
            shot_url = None
        if shot_url:
            img = os.path.join(base, f"{timestamp(f'{label}_{i}_')}.png")
            download_to_file(shot_url, img)
            return img
        return ""
    finally:
        _cleanup(client, c)


# --- ASYNC VERSION BELOW ---

async def _take_one_async(i: int, label: str, url: str, client: object | None = None) -> str:
    """Run _take_one in a separate thread so asyncio can manage concurrency."""
    return await asyncio.to_thread(_take_one, i, label, url, client)


async def run_async(n: int = 10, site: str | None = None, concurrency_limit: int = 5) -> List[str]:
    """Run up to n concurrent tasks using asyncio with a concurrency limit."""
    label, url = _select_site_arg(site)
    imgs: List[str] = []
    sem = asyncio.Semaphore(concurrency_limit)
    shared_client = Computer()

    async def worker(i: int):
        delay = 2.0
        attempts = 0
        while True:
            try:
                async with sem:
                    return await _take_one_async(i, label, url, shared_client)
            except Exception as e:
                # Print full traceback
                print(f"\n❌ Full error traceback for worker {i}:")
                traceback.print_exc()

                # Handle retry on capacity limit
                if _is_capacity_error(e) and attempts < 12:
                    attempts += 1
                    print(f"⚠️ Capacity limit; retrying iteration {i} after {delay:.1f}s (attempt {attempts}/12)")
                    await asyncio.sleep(delay)
                    delay = min(delay * 1.7, 60.0)
                    continue

                print(f"⚠️ Worker {i} failed permanently after {attempts} retries.")
                return ""

    # Run all workers concurrently, collecting results
    tasks = [worker(i) for i in range(n)]
    for coro in asyncio.as_completed(tasks):
        try:
            result = await coro
            if result:
                imgs.append(result)
        except Exception:
            print("\n❌ Unhandled top-level exception in a worker:")
            traceback.print_exc()
            # continue to next worker

    return imgs


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Take 10 screenshots concurrently using asyncio.")
    parser.add_argument("--site", help="Site label (wikipedia, nytimes, airbnb, github, reddit) or full URL", default=None)
    parser.add_argument("--n", type=int, help="Number of screenshots", default=10)
    parser.add_argument("--limit", type=int, help="Concurrent limit", default=5)
    args = parser.parse_args()

    imgs = asyncio.run(run_async(args.n, args.site, args.limit))
    for p in imgs:
        print(f"✅ Saved: {p}")
